### ISSUE TITLE
Focus

### ISSUE THREAD

**Issue by gvanrossum** [gvanrossum/patma#105]
I think we have one important problem to solve: load vs. store (#90). Everything else is easy and won’t be a blocker. The load vs. store issue also affects the wildcard spelling (#92). That’s it.

**Comment by brandtbucher** [comment]
I followed @viridia's lead and made a branch that refactors `expr.py` to use a suffix question mark for stores. I also went ahead and created constants for all of the string tokens.

I think the result looks really, really nice.  The question mark makes intuitive sense to me as a query. I want `x` and `y`, so I ask for them: `Point(x?, y?)`.  And I think it cleans up all of the major complaints that we saw on the Python-Dev thread: accessibility concerns, to foot-guns, etc:

https://github.com/gvanrossum/patma/blob/expr-qmark-suffix/examples/expr.py

I'm firmly in support of this, and less so in support of a `?` prefix. I think that reading/looking like English is important. I'm against allowing arbitrary expressions with `$()` or `+()` or something: it feels like *way* overkill with tons of subtle implications and edge-cases.

Perhaps the other authors can chime in and we can just settle this?

**Comment by viridia** [comment]
As I (just now) posted in the other thread, the only major obstacle to adopting an explicit symbol for stores is aesthetics. In every other respect I think that it is better, for all the reasons you mentioned.

There was one concern raised about compatibility with PEP 505 but this should not be an issue since it is a different parsing context.

My style of advocacy tends to eschew actually coming out in support of a position, I tend to take a more Socratic approach and merely point out the consequences of a decision, and let the other person decide what they want. But yes, if I had to give an opinion I would say I was in favor of the general approach, assuming a consensus on the look can be reached.

**Comment by viridia** [comment]
On the issue of prefix vs. suffix, the answer is 'it depends'.

If we choose '?' I can go either way. One advantage of choosing '?' over other symbols is that it also reads as 'wildcard' when used alone without a name. I admit that '?' when used as a prefix looks a bit odd, my only concern is that lots of languages use '?' as a suffix to mean different things, but that is only a minor objection. I don't feel super strong about this.

The next-most viable candidate is '$'. The reason for this is that `$` traditionally means 'interpolation' in many languages (bash, JavaScript, Perl, etc.) and named variables in patterns are kind of like the inverse of interpolation if you squint. In this case, it should definitely be a prefix. One argument against '$' is that it currently is not used in Python at all, and we should save it for something really special - cause once it's gone, there's no getting it back.

(I thought about proposing unary '%' as a prefix - based on the idea that in old-style Python format strings, '%' was the interpolation character. But no, that just looks too weird.)

'*' and ':' cannot be used because it would cause syntactical ambiguities.

I don't like two character sequences like `==`. Based on my sample code, named variables happen a _lot_, so I want to minimize the amount of clutter.

I could go through the rest of the ASCII character set, but nothing really jumps out at me as a viable alternative.

**Comment by JelleZijlstra** [comment]
I used `@` once in a toy implementation of pattern matching. The intuition is that in `case Point(@x, @y):`, the x coordinate goes "at" the variable `x`, and the `y` coordinate goes "at" y. Then again, `@` already has two other very different meanings in Python.

**Comment by Tobias-Kohn** [comment]
Going through the rest of the ASCII character set (at least a bit) does produce two other viable candidates: `@` (as proposed here by @JelleZijlstra) and `&` (as proposed by  [here](https://github.com/gvanrossum/patma/issues/90#issuecomment-650454582) by @dmoisset).

At least in _Scala_, the `@` is used analogously to our walrus operator, i.e. `x @ Pattern` in Scala means the same as `x := Pattern` means here.  I think I have seen it in other languages, too (I am not going through all of them to check again, though).  In Python, `@` is currently used as a magic marker for decorators and matrix multiplication.  Neither of this would clash with the new usage, and IMHO also not feel too alien.

I could probably also live with `&`, given its usage in C, or even `?` if that turns out to be the overwhelming concensus. However, from these choices I would definitely prefer `@`.  And the variant `x?` really looks more like load semantics to my eyes, as it seems to read 'is this `x`?'.  So, if `?`, then rather as a prefix than a suffix.

We should be aware, though, that as soon as introduce such a marker for store semantics, there will most probably be a discussion ensuing about `@x.attr`, `@x[0]`, etc.  Indeed, unpacking already allows to do things like `a[1].foo, a[0].bar = a[0].foo, a[1].bar`.  So, in the sense of @viridia's list #101, we might want to add as a general design principle that all bindings should be local, or accept that you could actually use any legal target, although we might not like it.

In that sense, choosing an explicit marker for either load or store semantics has ramifications beyond pure aesthetics.  With the _dot_-rule we had settled on before, it was already clear that `x.attr` could not act as a target because it was explicitly marked as a load.

In a way, the discussion about using a marker like `?` also nullifies a previous argument against using a rule whereas names with capital letters would have load semantics.  As I recall the main argument against this was that it would be a first in Python.  But, so would `?`.  Whatever we do: we end up introducing a new way of denoting load/store semantics.  Hence, should we also revisit the idea that names starting with a capital letter should be treated as constant values?

Finally, I don't think we should be too concerned with the question of whether any of these choices could act as a wildcard character (as I mentioned before, I really do not understand the problem with `_` in the first place).  Of course, the outcome of how we mark load/store semantics influences other aspects, such as the wildcard character.  But let's focus on the load/store semantics first.

**Comment by gvanrossum** [comment]
> as I mentioned before, I really do not understand the problem with `_` in the first place

Since `_` is a valid identifier, *if* we were to allow plain identifiers (without leading `.`) in constant value patterns, we'd have a problem if `_` was a wildcard.

PS. I will need to collect my thoughts after last week's wild ride before I can write coherently about this topic as a whole.

**Comment by viridia** [comment]
Initially I was attracted to the idea of using `?` as both a wildcard and as an explicit store marker purely out of a desire for parsimony - reducing the number of symbols required, bolstered by a somewhat tenuous argument that both concepts could be combined under a single larger concept.

However, given the fairly strong evidence that `_` is used as a wildcard in many other languages (thanks for doing that research @Tobias-Kohn !), I'm now fairly sold on the idea that we should leave the PEP as-is with respect to wildcards. As such, there is no need to tie together the two issues (marking stores and wildcards), and we can address the store issue independently, which simplifies things a bit.

At this point we have a bunch of contenders, here's the ones I have seen mentioned or can envision:
```
  # Implicit, with loads being explicitly marked
  # - it's the prettiest of the choices, given the high frequency of stores
  # - easy to accidentally overwrite a local name
  case Point(x, y):

  # Explicitly-declared by name via some to-be-determined syntax
  # - The most verbose of the proposals
  case Point(x, y) with x, y:

  # Prefix question mark
  # - question mark isn't often used as a prefix, looks weird to most people
  case Point(?x, ?y):

  # Suffix question mark
  # - kind of looks like the start of a C/Java ternary operator, or null-aware member access
  case Point(x?, y?):

  # Dollar sign
  # - inspired by Unix shell, make, string interpolation
  case Point($x, $y):

  # At-sign
  case Point(@x, @y):

  # Ampersand 
  # - from C/C++ variable reference
  case Point(&x, &y):

  # Percent 
  # - from old-style Python string templating
  case Point(%x, %y):

  # Backticks
  case Point(`x`, `y`):

  # Angle brackets - from C++/Java templates
  case Point(<x>, <y>):
```
I'm sure that someone can make up an aesthetic rationalization for each of these choices, the question is how do we decide between them, especially given that no two people have the same aesthetic sensibilities?

The consequence of choosing one of these alternatives is mostly aesthetic, but not entirely:

* Some choices are predictably more likely to generate coding errors than others.
* Using a symbol which has never been used in Python before has advantages, in the sense that users won't be confused by the different meanings in different contexts, but it also has an opportunity cost, since it means that symbol is now no longer available for other uses in the future.

At this point, I don't think we are going to get a unanimous consensus on any of these choices, so the question is how to we get to a "least disliked" result?

Getting any kind of objective data to make a decision would be hard - whether that means polling Python users or doing some kind of end-user testing.

I suppose we could start by winnowing down this list and getting consensus on identifying the ones that we definitely *don't* want - but then the question is how to make that decision stick, when there is so much creative energy around this PEP.

**Comment by gvanrossum** [comment]
Since we all agree that stores are (much) more common than loads, why are all the contenders making stores uglier? Why not make loads uglier? Just because the one thing we tried (prefix `.` if there's no internal `.`) didn't fly?

I'd like to try a different line of reasoning. Please bear with me, this is not a polished argument (yet).

There are two ways to tell the story of `match`. One story starts with `switch` on steroids -- `case 1`, `case (1, 2)`, `case {"x": 1, "y": 2}`, `case Point(1, 2)`, a way to combine values (`1|2`) and of course nested values like `Rect(Point(...), Point(...))` -- and then adds wildcards (`case Point(1, _)`) and finally capture (`case Point(x, y)`). This story is derailed because `1` and `2` are begging to be generalized to arbitrary expressions, or at least named constants -- but then how to spell capture? This seems to be where we're stuck, exploring various ways of spelling capture in a way that does not confuse it with (a limited form of) expressions.

The other story may be more compelling, and this is how I was introduced to `match` in Scala. It primarily sees `match` as a switch on the *type* (or sometimes shape) of an object. It starts with `match` probing different types/shapes in turn, similarly to tuple unpacking: `case (x, y)`, `case (x, y, z)`. Then it generalizes to algebraic types: `case Point(x, y)`. Wildcards fall naturally out of the existing use of `_` as a throwaway. Next we observe that sometimes we want to treat single values as "types", similar to how PEP 586 uses `Literal[...]`. Since at this point in the story, patterns are built out or variables (`x`, `y`) and syntax denoting structure (`(...)`, `Point(...)`), we can conveniently use literals (`1`, `2`, ...) to stand for themselves. But now the story derails if we want to support named constants.

Finally, I would like to observe that the generalization to *arbitrary* expressions is doomed either way -- it leads to confused users trying
```
pxy = Point(x, y)
match target:
    case pxy:
        print(x, y)
```

Note that Scala has exactly this problem: https://scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#stable-identifier-patterns (A stable identifier is a effectively named constant.)

> To resolve the syntactic overlap with a variable pattern, a stable identifier pattern may not be a simple name _starting with a lower-case letter_. However, it is possible to enclose such a variable name in backquotes; then it is treated as a stable identifier pattern. [Italics mine]

**Comment by gvanrossum** [comment]
PS. It seems elixir uses a `^` prefix (which is calls "pin") to indicate a load instead of a store: https://elixir-lang.org/getting-started/pattern-matching.html#the-pin-operator

**Comment by viridia** [comment]
> Why not make loads uglier?

The issue that was brought up by several people on python-dev is that, in a nutshell, stores are "more dangerous" than loads, and it's possible that a programmer may make a mistake and indicate a load when they meant store and vice versa.

There are two possible errors to consider: one is accidentally omitting the marker, and the other is accidentally including it. Since the whole purpose of the marker is to stand out visually, I think we can agree that the former case is much more likely than the latter; moreover, the latter bug will be much easier to spot and fix (and uglier markers will tend to stand out more). So let's just consider the consequence of omitting the marker.

If loads require a marker, and stores do not, then the consequences of accidentally omitting the marker converts a load into a store - meaning that the pattern will overwrite the variable. Since (a) name patterns always match, and (b) the name used by the programmer will always be a pre-existing symbol (why bother to load a non-existing symbol), then the result of the error will be to overwrite that name. This is a runtime error that might not be detected immediately.

If stores require a marker, and loads do not, then the consequence of accidentally omitting the marker converts a store into a load. This will be a compilation error in the vast majority of cases - because most stores are referencing a name that was not previously defined in the local scope (because otherwise you'd be shadowing a variable, which would be considered a code smell).

So in other words, the mistake of leaving out the marker is either a sleeping time bomb in your code, or an immediate compilation error.

None of this matters if you don't make a mistake in your code. So the issue to decide is whether the aesthetic considerations outweigh the possible prevention of typos. Because this is a value judgement, it is clear that some readers weigh the relative impact of these concerns differently. Some readers will think that errors of this kind are unlikely in the first place and therefore it is not a concern, while others experience a heightened degree of anxiety thinking about the potential mistakes. That is, I believe, the source of the contention.

**Comment by Tobias-Kohn** [comment]
These are really great posts about the story/usage and the issue of possible errors.  It seems we are actually discussing along two different, somewhat orthogonal dimensions: _usage frequency_ vs. _error probability_.  Let me perhaps build on Guido's post about the story behind pattern matching here...

Here is a radical thought: what if we completely concentrated on the core of pattern matching as deconstruction of complex data values, with no load semantics altogether?  We could put the idea of load semantics to the deferred section.  This is probably not the most popular choice among all those who see this as a nice way to finally have `switch` in Python.  But it would make it quite clear that this is actually not about having `switch`, but really about types and structure of data.  And since there is no load semantics altogether, there is no foot-gun, although we cannot prevent all errors and mistakes.

---

In a way, the problem also stems from Python not having constants (I am not saying that it should, though).  If there was a way to tell Python that `PIE`, say, is a fixed constant, it becomes immediately clear to the compiler that it cannot have store semantics (since it is a constant), and it can thus fall back to load semantics, instead.  The only constants as such that do exist are literal values, which we readily support, anyway.

The issue with not having constants is also a major obstacle for having a `switch` statement as such.  One of the primary appeals of `switch` is the idea of direct and fast dispatch, usually thanks to a jump table.  But as soo as you use anything other than literal values, that jump table would not be static, but would have to be rebuilt basically for each iteration, nullifying its advantage.  So, you need to have an `if`-`elif`-`else` chain underneath, anyway.

By favouring load semantics, we basically accept that the new feature is foremost a `switch` on steroids with its primary use case being to replace `if`-`elif`-`else` chains.  This somewhat contradicts the beginning of the PEP, where Ivan pointed out that `isinstance()` is one of the most often called functions, implicitly stating that this is what we want to address.  And for a `switch` on steroids, we probably would want to have something like range-patterns with rather high priority.

Finally, let's also keep in mind that the load semantics is actually a _load and compare_ semantics.  The 'and compare' part comes with its own set of difficulties, as we have already discussed in issue #16.  But there is also a [discussion on the Haskell mailing list](https://www.mail-archive.com/haskell-cafe@haskell.org/msg59618.html) that points out why patterns like `case (x, x)` are problematic: in general, it is undecidable whether a given pattern has such a structure&mdash;although Python is rather pragmatic here with a well defined semantics for `==` comparison, which is allowed to err (`==` can fail on both sides in not discovering a difference, as well as not seeing that two things are equal).

**Comment by gvanrossum** [comment]
Both great posts. I had been thinking along the same lines (remove loads altogether) and hesitated, because naming constants is culturally ingrained. But if the only way (or the most convenient way) to save clean store semantics is to drop load semantics altogether I'm for it.

And yes, it could be deferred -- in the future we could add a `^` prefix or use a pair of backticks or go back to the dotted-names rule. We couldn't easily introduce Scala's Uppercase convention unless we restricted the variable names to start with a lowercase letter in the current proposal though.

PS. I wish we had this discussion in #90, but here we are.

**Comment by brandtbucher** [comment]
Let’s also not forget that all loads can be accomplished with stores + guards. Consider the following example from the PEP:

```py
match color:
    case .BLACK | Color.BLACK:
        print("Black suits every color")
    case BLACK:  # This will just assign a new value to BLACK.
        ...
```

When outlawing all name loads (except class names, of course), this becomes:

```py
match color:
    case c if c in (.BLACK, Color.BLACK):
        print("Black suits every color")
    case BLACK:  # This will just assign a new value to BLACK.
        ...
```

I’m not sure if I like this or not. It is annoying to those wanting a `switch` statement, but it drastically simplifies the rules users need to remember here. Plus, maybe discouraging switching is a good thing (`if`/`elif`/`else` will always be the obvious way to do it).

**Comment by gvanrossum** [comment]
Maybe it'll encourage a new style of defining enums, as algebraic types. Then you could write
```
match color:
    case RED():
        print("You'll stand out in any crowd")
    case BLACK():
        print("Black suits are always in style")
```
I could imagine some kind of alternate Enum class that would let you write
```
class Color(BetterEnum):
    RED = 0
    BLACK = 1
```
and behind the scenes it would create values that are subtypes instead of instances of `Color`.

It would make enums expensive though -- each enum value would be its own class, and class objects have a huge amount of overhead.

**Comment by viridia** [comment]
I would hate to drop loads, especially in sub-patterns. The classic 'visitor pattern' involves both destructuring and testing against an ID:

```
  case ASTNode(PLUS, arg1, arg2):
  # dozens more similar cases follow
```
Using guards this becomes more cumbersome, less readable and arguably less performant:
```
  case ASTNode(op, arg1, arg2) if op == PLUS:
```

**Comment by dmoisset** [comment]
I tried an alternate take for this and some other things in #116 ; starting with the idea that Enum's are classes are both somewhat different forms of algebraic types, and trying to unify from that. The result is not obvious, but I think it provides at least a coherent story, adding a "Constructor Pattern"which is essentially the class pattern with slightly extended semantics (replacing it and also the constant value pattern)

**Comment by gvanrossum** [comment]
I think we can close this. Let's open new issues as they come up in the feedback over the next few days.
